File: watstats.pl
#!/usr/bin/env  perl
# compute some statistics for shotgun projects
# $Id: watstats,v 1.2 2002/12/06 11:21:49 fm Exp fm $
# Programming by Pierre Rioux, OGMP, 1994
# Fixing by Folker Meyer, Bielefeld University, 2001

use Getopt::Std;
              
################################################################
# read arguments
getopts('hvg:n:l:t:');

if (defined $opt_h) {
	&usage;
	die "\n";;
}

if (defined $opt_v){
 $debug=1;
}

# number of reads (default 500)
if (defined $opt_n){
  if ($debug) {
     print "$0 Using N:$opt_n \n";
  }
  $N=$opt_n;
}
else {
 $N=500;
}

# genome size in kb (default = 70)
if (defined $opt_g){
  if ($debug) {
     print "$0 Using G:$opt_g \n";
  }
  $G=$opt_g;
}
else {
 $G=70
}

# avg read length (default = 480)
if (defined $opt_l){
  if ($debug) {
     print "$0 Using L:$opt_l \n";
  }
  $L=$opt_l;
}
else {
 $L=480
}

# overlap length (default = 30)
if (defined $opt_t){
  if ($debug) {
     print "$0 Using T:$opt_t \n";
  }
  $T=$opt_t;
}
else {
 $T=30
}
# parsing done
##########################################

$O=200;

$TABLE=0;
$OCEAN=0;

    $VV="";
    foreach $n ("G","L","N","T","O") {
        eval "(\$x,\$y,\$z) = (\$$n =~ /^\\D*(\\d+)\\D+(\\d+)\\D+(\\d+)/);";
        next if $x eq "";
        if ($VV) {
            print "Error: you can't make more than one variable parameter.\n";
            $VV="ERROR";
            last;
            }
        ($from,$to,$step) = ($x,$y,$z);
        $VV=$n;
        eval "\$$VV=$from;";
        $TABLE=1;
        }
    next if ($VV eq "ERROR");

    if (!$VV) {
        $VV="G";
        $from=$G;$to=$G;$step=1;
        $TABLE=0;
        }

    next if ($G <= 0 || $T <= 0 || $L <= 0 || $N <= 0 || $O <= 0);

    if ($TABLE) {
        eval "\$$VV=\"From $from to $to, step of $step\"";
        }

    print "Length of genome            G = ${G}000 bp\n";
    print "Number of sequences         N = $N\n";
    print "Average length of sequences L = $L bp\n";
    print "Minimum overlap             T = $T bp\n";
    print "Ocean threshold             O = $O bp\n" if $OCEAN;

    if ($TABLE) {
        print "\n";
        print "$VV\tCover\t#Contig\t#Seq/C\tContLen\t%Compl";
        print "\tOcean>$O" if $OCEAN;
        print "\n-\t-----\t-------\t------\t-------\t------";
        print "\t---------" if $OCEAN;
        print "\n";
        }

    for ($range=$from;$range<=$to;$range+=$step) {
        eval "\$$VV=\$range;";

        $GM = $G*1000;
        $alpha=$N/$GM;
        $theta=$T/$L;
        $sigma=1-$theta;
        $c    =$L*$N/$GM;

        #$sigma=1;  # Override: means overlap ALWAYS detected.

        $cover=$c;
        $i  =$N*exp(-$c*$sigma);                            $i=1    if $i   < 1; $i  =$N if $i   > $N;
        $iii=exp($c*$sigma);                                $iii=$N if $iii > $N;$iii=1  if $iii < 1;
        $iv=int($L*(((exp($c*$sigma)-1)/$c)+(1-$sigma)));   $iv=$GM if $iv  > $GM;
        $vi=int(100*exp(-$c*($O/$L+$theta)));
        $compl=int(100*$i*$iv/$GM);

        &AdjustValue(*i);
        &AdjustValue(*iii);
        #&AdjustValue(*iv);
        &AdjustValue(*cover);

        if (!$TABLE) {
            print "\n";
            print "Redundancy of coverage                   = $cover\n";
            print "Expected number of contigs               = $i\n";
            print "The expected number of seqs in a contig  = $iii\n";
            print "The expected length of each contig       = $iv\n";
            print "Percentage of coverage of genome         = $compl%\n";
            printf"The prob of having an ocean > %s bp     = %-3.1f%%\n",$O,$vi
                if $OCEAN;
            }
        else {
            if ($OCEAN) {
                print "$range\t$cover\t$i\t$iii\t$iv\t$compl%\t";printf "%-3.1f\n",$vi;
                }
            else {
                print "$range\t$cover\t$i\t$iii\t$iv\t$compl%\n";
                }
            }
        }
        if ($TABLE) {
            eval "\$$VV=\"From $from to $to, step of $step\"";
            }


#########################################################
sub usage{
print <<BANNER;
USAGE: watstats -g GENOMESIZE in KB 
                -n # of reads (either number or from-to,step)
                -l avg. read length
                -t min. overlap needed

example:  watstats -g 5000 -n 1000-70000,500 -l 480 -t 30

watstats 1.1 ; Based on

    Lander, E. S. and Waterman, M. S., "Genomic mapping by fingerprinting
    random clones: a mathematical analysis", Genomics 2, 231-239 (1988).

Programming by Pierre Rioux, OGMP, 1994
Fixing by Folker Meyer, ZfGB, 2001
BANNER
}
##################################
sub AdjustValue {
    local(*VarName) = @_;
    $VarName=sprintf("%4.1f",$VarName);
}

File: watstats.r
watstats <- function (
                      num_reads = 1000000,
                      percent_data = 20, # use to get num reads
                      genome_length = 4000000,
                      read_length = 125,
                      min_overlap = 30,
                      verbose=TRUE
                      )
  {

     taxa_num_reads = ( num_reads * ( percent_data/100 ) ) # 
     
     alpha    <<- ( taxa_num_reads/genome_length ) # $alpha=$N/$GM; # $GM = $G*1000; (input in original was in KB)
     theta    <<- ( min_overlap/read_length )      # $theta=$T/$L;
     sigma    <<- ( 1-theta )                      # $sigma=1-$theta;

     coverage_redundancy <<- ( (read_length*taxa_num_reads)/genome_length )          # $c=$L*$N/$GM;

     num_contigs      <<- taxa_num_reads*exp(-coverage_redundancy*sigma)    # $i  =$N*exp(-$c*$sigma); 
     if ( num_contigs < 1 ){ num_contigs <<- 1 }                            # $i=1    if $i   < 1;
     if ( num_contigs > num_reads ){ num_contigs <<- num_reads }            # $i  =$N if $i   > $N;
     
     seqs_per_contig  <<- exp(coverage_redundancy*sigma)                             # exp($c*$sigma);
     if ( seqs_per_contig > num_reads ){ seqs_per_contig <<- num_reads }             # $iii=$N if $iii > $N;
     if ( seqs_per_contig < 1 ){ seqs_per_contig <<- 1 }                             # $iii=1  if $iii < 1;

     contig_length    <<- read_length*(((exp(coverage_redundancy*sigma)-1)/coverage_redundancy)+(1-sigma)) # $iv=int($L*(((exp($c*$sigma)-1)/$c)+(1-$sigma)));
     if ( contig_length > genome_length ){ contig_length <<- genome_length }         # $iv=$GM if $iv  > $GM;
  
     percent_coverage <<- 100*num_contigs*contig_length/genome_length                # $compl=int(100*$i*$iv/$GM);
     if ( percent_coverage > 100 ){ percent_coverage <- 100 }
     
     if( verbose==TRUE ){
       print("INPUT")
       print(paste("num_reads           :", round(taxa_num_reads, digits=0)))
       print(paste("percent_data        :", round(percent_data, digits=1)))
       print(paste("genome_length       :", round(genome_length, digits=0)))
       print(paste("read_length         :", round(read_length,digits=0)))
       print(paste("min_overlap         :", round(min_overlap, digits=0)))
       print("")
       print("OUTPUT")
       print(paste("coverage_redundancy :", round(coverage_redundancy, digits=1)))
       print(paste("num_contigs         :", round(num_contigs, digits=1)))
       print(paste("seqs_per_contig     :", round(seqs_per_contig, digits=1)))
       print(paste("contig_length       :", round(contig_length, digits=1)))
       print(paste("percent_coverage    :", round(percent_coverage, digits=1), "%"))
       print('------------------------------------------------------')
     }
     
     watstats_command <- paste("watstats.pl -g", genome_length, "-n", taxa_num_reads, "-l", read_length, "-t",  min_overlap, "intern=TRUE")
     
     my_watstats <<- system(watstats_command)

    #watstats_command <- paste("watstats.pl -g", genome_size, "-n", taxa_num_reads, "-l", read_length, "-t",  min_overlap, "intern=TRUE")

    # system2(watstats_command, stdout=my_watstats.2)
    
    #my_watstats.2 <<- capture.output(system(watstats_command), file = NULL, append = FALSE)
    
    my_date <<- system("date", intern=TRUE)


  }
# watstats.pl -g 5000 -n 1000 -l 480 -t 30

# watstats 1.1 ; Based on
#    Lander, E. S. and Waterman, M. S., "Genomic mapping by fingerprinting
#    random clones: a mathematical analysis", Genomics 2, 231-239 (1988).
